Developer Tutorial Move - 2 (Command Local)

kb1lqd opened the pull request:
# Pull Request Overview

This pull request updates the `Commanding-Local` tutorial script and readme.md.

## Notable Updates
* Spellchecking and other typo's fixed is both readme and example script
* General wording + partial PEP 8 update (easy stuff)
* Updated script to use an INI file for configuration
* Added exception handling to ECHO command for both good practice and due to noted bug ( #50 ) 
  * Added section in readme about the workaround

## Testing Notes

* I've verified it works as written in readme and intended by script

Commit:
Initial Copy - Command Local

Commit:
Fixed basic Readme.md Typo's

Commit:
Updating wording in readme halfway

Commit:
Fixed wording.

Commit:
Merge remote-tracking branch 'refs/remotes/origin/master' into Developer_Tutorial_Dev_Move_2

Commit:
Merge remote-tracking branch 'refs/remotes/origin/master' into Developer_Tutorial_Dev_Move_2

Commit:
PEP8 Quick + minor fixes in script

Commit:
Reviewed and updated wording in script

Commit:
Added ECHO exception

Added execption due to proxy not responding to first packet bug. Also
good practice when potential for data to not be present.

Commit:
Added ECHO exception in readme.md

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py:
```diff
@@ -1,87 +1,110 @@
+                                   "UNIT0CALL")  # Should match the connected Faraday unit as assigned in Proxy configuration
+local_device_node_id = config.get("DEVICES",
+                                  "UNIT0ID")  # Should match the connected Faraday unit as assigned in Proxy configuration
+local_device_callsign = str(local_device_callsign).upper()
```
config.get() is already a string and just perform .upper() to it on the same line.

kb1lqd replied:
Yep, completed.

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py:
```diff
@@ -1,87 +1,110 @@
 ##############
-#Turn LED 1 ON LOCAL
+# Turn LED 1 ON LOCAL
 print "Turning ON the Green LED (LED #1)"
```
I won't block on it but we should really use logging instead of print.

kb1lqd replied:
Well I agree our applications and larger tutorials should be upgraded to logging this is simply actively telling the user when the LED should be ON. Print is appropriate.

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py:
```diff
@@ -1,87 +1,110 @@
+# Variables
+local_device_callsign = config.get("DEVICES",
+                                   "UNIT0CALL")  # Should match the connected Faraday unit as assigned in Proxy configuration
+local_device_node_id = config.get("DEVICES",
```
might want to use .getint()

https://docs.python.org/2/library/configparser.html#ConfigParser.RawConfigParser.getint

kb1lqd replied:
Updated node ID to gitint() but need to verify working.

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py:
```diff
@@ -1,87 +1,110 @@
+# Turn LED 2 OFF LOCAL
 print "Turning Off the Red LED (LED #2)"
-command = faraday_cmd.CommandLocalGPIO(0, 0, 0, gpioallocations.LED_2, 0, 0) #This examples how the non predefined LED GPIO commanding is created. Multiple GPIO's can be toggled at once using ||'s
+command = faraday_cmd.CommandLocalGPIO(0, 0, 0, gpioallocations.LED_2, 0, 0)  # This examples how the non predefined LED GPIO commanding is created.
```
wait why not use `CommandLocalGPIOLED2Off()?` like you used for LED1 since you used a similar command for LED1? Based on inline comment that's the goal?

kb1lqd replied:
Correct, the inline comment explains the intent to show the raw bitmask function as well.

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py:
```diff
@@ -1,87 +1,110 @@
+    echo_decoded = faraday_1.DecodeRawPacket(b64_data)
+
+    # Display information
+    print "**Sending**\n"
```
I feel like this should print prior to actually sending the data

kb1lqd replied:
I moved the "sending..." print to above receiving code and after initializing the message.

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py:
```diff
@@ -1,87 +1,110 @@
+    print "\nRAW Received BASE64 ECHO'd Message:", b64_data
+    print "\nDecoded BASE64 RAW Bytes:", repr(echo_decoded)
+
+except Exception as e:
```
Broad exceptions like this aren't very good. Can let this one pass I guess but we should also try/except on specific lines we know what will fail and how it will fail. Those print statements are unlikely to cause problems we would not already have caught elsewhere.

kb1lqd replied:
I updated to Type Error, I think this is correct? "except TypeError as e"

kb1lqc replied:
Well it depends.

If line 91 fails it could be an IndexError or KeyError
Like 92 and the print statements are likely just covered by a StandardError. The trick is you want to react accordingly and appropriately. While catching the exception can prevent program crashing you could insert valid but filler data or rollback to a safe state depending on the error and circumstances.

Instead of just having a KeyError for `b64_data = rx_echo_raw[0]['data']` there could be a common issue with Proxy where you can inform the user to check a specific input value or configuration.

kb1lqc replied:
First time I ran this program after starting Proxy:
![image](https://cloud.githubusercontent.com/assets/331540/23397527/c5e13216-fd4d-11e6-944e-23aeacf6db70.png)

Second time:
![image](https://cloud.githubusercontent.com/assets/331540/23397537/d05558f8-fd4d-11e6-9204-e121a9a37c0c.png)

---
Yes if I restart Proxy I get the same error. Looking at Proxy the following error occurs the first time this program is run after starting Proxy:
![image](https://cloud.githubusercontent.com/assets/331540/23397605/1ce0b852-fd4e-11e6-9323-cb778c5066f8.png)

What's up with this? I've confirmed this exists but also, you've sent commands to port 2 for LED's. I do not understand why only ECHO fails on the first try.

I printed out the rx_exho_raw and you are getting a JSON error message back. I reccommend using that error message in your exception code:
![image](https://cloud.githubusercontent.com/assets/331540/23397978/84b47c74-fd4f-11e6-9b3d-bc4307b8236b.png)

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/configuration-template.ini:
```diff
@@ -0,0 +1,4 @@
+[DEVICES]
+UNITS=1
+UNIT0CALL=REPLACEME
+UNIT0ID= REPLACEME
```
Remove space after `=`

kb1lqd replied:
Done

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -1,29 +1,33 @@
-Following the [Configuring Proxy](../../0-Welcome_To_Faraday/Configuring_Proxy/) tutorial configure, start, and ensure a successful connection to a locally (USB) connected Faraday digital radio.
+## Configuration
 
+* Open `configuration-template.ini` with a text editor
```
I would use the `xxxxx.sample.ini` convention proposed in #110

kb1lqd replied:
Yeah I think I did the new file + .gitignore but not sure, can you check?

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -1,29 +1,33 @@
+## Configuration
 
+* Open `configuration-template.ini` with a text editor
+* Update `CALLSIGN` Replace `REPLACEME` to match the callsign of the Faraday unit **as assigned** in proxy
```
Funky wording
>Update `CALLSIGN` Replace `REPLACEME` to match

Update `CALLSIGN` from `REPLACEME` to match the callsign....

kb1lqd replied:
Fixed

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -1,29 +1,33 @@
 
+* Open `configuration-template.ini` with a text editor
+* Update `CALLSIGN` Replace `REPLACEME` to match the callsign of the Faraday unit **as assigned** in proxy
+* Update `NODEID` to match the callsign node ID of the Faraday unit **as assigned** in proxy
```
Probably the same as `CALLSIGN` as I suggested above

kb1lqd replied:
Yep

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -1,29 +1,33 @@
-> Note: GPIO's being commanded may be recieving other commands and not work as intended if multiple programs are controlling a single GPIO (i.e. RED due to RF TX indication)
+## Tutorial Output Examples
+
+below is a screen-shot of the partial output of the tutorial script when run in a python interpreter (PyCharm). Be sure to look at the two LED's on Faraday as the script is running to observe them turing ON and OFF.
```
Spelling "turning"

kb1lqd replied:
Fixed

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -33,7 +37,7 @@ below is a screen-shot of the partial output of the tutorial script when run in
 ## Code - Toggle GPIO's (LEDs) Predefined Functions
 
-The Faraday command module object pre-defines many common actions such as turning ON/OFF the on-board LED's. Using `faraday_cmd.CommandLocalGPIOLED1On()` will command the LED #1 to an ON state and light up the LED. Alternatively `faraday_cmd.CommandLocalGPIOLED1Off()` turns the LED OFF.
+The Faraday command module object predefine's many common actions such as turning ON/OFF the on-board LED's. Using `faraday_cmd.CommandLocalGPIOLED1On()` will command the LED #1 to an ON state and light up the LED. Alternatively `faraday_cmd.CommandLocalGPIOLED1Off()` turns the LED OFF.
```
predefine's is wrong should be predefines

kb1lqc replied:
You also refer to `LED #1` here but `LED#1` earlier in the document

kb1lqd replied:
Where?

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -63,14 +67,14 @@ time.sleep(1)
 ## Code - Toggle GPIO's (LEDs) Bitmask
 
-The previous code used pre-defined functions for toggling GPIO and left LED #2 ON. Those predefined functions are actually just providing an abstraction of the raw command for GPIO bitmask control. The GPIO bitmask control is part of the Command Application and provides both ON/OFF control of many GPIO in a single command while also guarding potentially damaging GPIO from being accidently changed ...
+The predefined functions are actually just providing an abstraction of the `CommandLocalGPIO(self, p3_bitmask_on, p4_bitmask_on, p5_bitmask_on, p3_bitmask_off, p4_bitmask_off, p5_bitmask_off)` function that provides GPIO bitmask control of all controllable GPIO pins. This functionality is part of the Command Application and provides control of all GPIO in a single command while also guarding pote ...
```
> ...while also guarding potentially damaging GPIO from being accidentally changed
Weird wording

...while also guarding against changing GPIO values which could damaging a GPIO pin.

kb1lqd replied:
Updated.

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -106,14 +110,15 @@ faraday_1.POST(local_device_callsign, local_device_node_id, faraday_1.CMD_UART_P
+A simple ECHO like command is provided that will send the supplied data payload back to the host computer. The ECHO command only accepts a single UART transport packet so it is limited in size to the 124 byte maximum transmissible unit of the current UART network stack.
 
-This command example is pretty self explaining and simply sends the string `"This will ECHO back on UART"` to the local Faraday device and retrieves, parses, and displays the ECHO'd packet. Note the padding bytes likely visible due to fixed length UART transport packet formatting appended to the original message. Also, this command accepts binary data values (not just ASCII string text)!
+This example sends the string `"This will ECHO back on UART"` to the local Faraday device and retrieves, parses, and displays the ECHO'd packet. Note the appended padding bytes visible due to fixed length UART transport packet.
```
Better wording

...to the local Faraday device retrieving, parsing, and displaying the ECHO'd packet.

kb1lqd replied:
Fixed

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -143,15 +150,18 @@ below is a screenshot of the partial output of the tutorial script when run in a
+#Bonus Exercise
+
+* Modify the the example script to remove the padding bytes and display only the original ECHO'd string.
+  * Can you support variably given any string length using only the packets sent and received ("Decoded BASE64 RAW Bytes")?
```
What? Variably? Sentence does not make sense to me.

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -143,15 +150,18 @@ below is a screenshot of the partial output of the tutorial script when run in a
+* Modify the the example script to remove the padding bytes and display only the original ECHO'd string.
+  * Can you support variably given any string length using only the packets sent and received ("Decoded BASE64 RAW Bytes")?
 
+# To-Do
```
Why have todo's here?

kb1lqd replied:
Ok, I'll remove.

kb1lqc reviewed (changes requested):
Haven't yet run through using hardware but found enough spelling and wording stuff to have you get that working while I work on using this with hardware.

Commit:
Added gitint(), need to verify working

Commit:
Finished general update to script

Commit:
Updated INI to new .sample.ini

Commit:
Updated readme.md with .sample INI

Commit:
Update git ignore for developer turorial

Commit:
Reverted to general exception

Commit:
Updated typo's and wording from final review push

Commit:
Removed to-do

kb1lqd commented:
@kb1lqc Please re-review and note my comments with questions/checks that I did things correctly. Mostly, did I update the .gitignore and rename the INI correctly.

Commit:
Updated main tutorial readme to remove outdated note

kb1lqc commented on Tutorials/Python_Developer_Tutorials/Readme.md:
```diff
@@ -19,7 +19,7 @@ Each tutorial series builds upon the prior to create a simple wireless transmiss
 ##Introduction To Faraday Proxy Programming
 
-* **[Commanding - Local](foundation/Commanding-Local/) (OUTDATED - UPDATE IN PROGRESS)**
+* **[Commanding - Local](foundation/Commanding-Local/)
```
Stray "**" here

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -1,29 +1,33 @@
-below is a screen-shot of the partial output of the tutorial script when run in a python interpreter (PyCharm). Be sure to look at the two LED's on Faraday as the script is runing to observe them turing ON and off.
 
-> Note: GPIO's being commanded may be recieving other commands and not work as intended if multiple programs are controlling a single GPIO (i.e. RED due to RF TX indication)
+## Tutorial Output Examples
```
You never actually tell the user how to run the program!

kb1lqc reviewed (changes requested):
Some errors occurred when running. It appears that ECHO function acts differently than the LED functions causing the problems. I do not expect this:
![image](https://cloud.githubusercontent.com/assets/331540/23398021/bf211b88-fd4f-11e6-968a-80e5480c7287.png)

Since LED commands were already sent several times prior, the port has been used!

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py:
```diff
@@ -1,87 +1,110 @@
-
-
+#  Retrive waiting data packet in UART Transport service number for the COMMAND application (Use GETWait() to block until ready or timeout
+rx_echo_raw = faraday_1.GETWait(local_device_callsign, local_device_node_id, faraday_1.CMD_UART_PORT,
```
This should be checked for the "error" JSON response as shown below:
![image](https://cloud.githubusercontent.com/assets/331540/23398334/fd3c5fe4-fd50-11e6-85fe-8a1aee975197.png)

failing gracefully and printing the error message returned by proxy is what we should do. This is necessary in all applications but just noticed that we weren't using this functionality I built in to the API.

This is failing on first try because the GETWait() because you are trying to obtain data from the CMD_UART_PORT for the first time. All prior LED operations were a POST() to the port. This means that the buffer hasn't been created yet for this particular port on the RX side of things. This is Issue #60 creeping up.

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py:
```diff
@@ -1,87 +1,112 @@
+    print "\nDecoded BASE64 RAW Bytes:", repr(echo_decoded)
+
+except Exception as e:
+    print "Failed ECHO due to type error:", e
```
Not really a type error. Would be MOST useful to print out the error JSON message here.

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -143,15 +150,13 @@ below is a screenshot of the partial output of the tutorial script when run in a
-* Modify the the example script to remove the padding bytes and display only the original ECHO'd string.
-* Can you do this variably given any string length using only the packets sent and received?
-  * *Hint: [Application layer](https://en.wikipedia.org/wiki/OSI_model) packet [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(networking))*
+There is a [known bug](https://github.com/FaradayRF/Faraday-Software/issues/50) that may cause the first packet from Proxy to not properly be retrieved and this has caused ECHO to fail. The workaround to this until it is patched is to simply re-run the tutorial script **without** restarting proxy.
```
Ahh I missed that you noted this already. It's issue #60 not 50

Commit:
Added how to run tutorial and moved bug note to top of document

Commit:
Changed exception to KeyError as it should be instead of generic error and added more detailed description to error print statement

Commit:
Reseting proxy.ini back to master manually...

kb1lqd commented:
@kb1lqc I updated per comments again:

* Added section to instruct how to run tutorial script
* Moved known bug description/info to top of readme in tutorial script running section
* Changed general exception to "KeyError" as expected by known bug error. I also added a print statement to describe the bug/workaround

kb1lqc committed:
Removed extra **

kb1lqc committed:
Spelling and formatting fixes

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -63,14 +79,14 @@ time.sleep(1)
 ## Code - Toggle GPIO's (LEDs) Bitmask
 
-The previous code used pre-defined functions for toggling GPIO and left LED #2 ON. Those predefined functions are actually just providing an abstraction of the raw command for GPIO bitmask control. The GPIO bitmask control is part of the Command Application and provides both ON/OFF control of many GPIO in a single command while also guarding potentially damaging GPIO from being accidently changed ...
+The predefined functions are actually just providing an abstraction of the `CommandLocalGPIO(self, p3_bitmask_on, p4_bitmask_on, p5_bitmask_on, p3_bitmask_off, p4_bitmask_off, p5_bitmask_off)` function that provides GPIO bitmask control of all controllable GPIO pins. This functionality is part of the Command Application and provides control of all GPIO in a single command while also guarding agai ...
```
Since when did we get a "Command Application"?

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -63,14 +79,14 @@ time.sleep(1)
 * 3 function arguments for commanding GPIO ports 3, 4, 5 pins OFF
 
-If any bits are HIGH (1) in the bitmask bytes then that respective pin will be toggled ON (HIGH), if a bit is left LOW (0) then no action will take place and the bit will remain in it's current HIGH/LOW state.
+If any bits are HIGH (1) in the bitmask bytes then the respective ON/OFF action will be applied to that bitmask's pin, if a bit is left LOW (0) then no action will take place and the pin will remain in it's current HIGH/LOW state. Commanding both and HIGH and LOW for the same port/pin is not allowed.
```
what happens when you command both ON and OFF at the same time?

kb1lqd commented:
The python module function won’t allow it and will fail and should return False IIRC.

 

From: Bryce Salmi [mailto:notifications@github.com] 
Sent: Thursday, March 2, 2017 11:19 PM
To: FaradayRF/Faraday-Software <Faraday-Software@noreply.github.com>
Cc: Brenton Salmi <kb1lqd@gmail.com>; Assign <assign@noreply.github.com>
Subject: Re: [FaradayRF/Faraday-Software] Developer Tutorial Move - 2 (Command Local) (#111)

 

@kb1lqc commented on this pull request.

  _____  

In Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md <https://github.com/FaradayRF/Faraday-Software/pull/111#discussion_r104100181> :

>  
 The Python module function `faraday_cmd.CommandLocalGPIO()` contains 6 function arguments:
 
 * 3 function arguments for commanding GPIO ports 3, 4, 5 pins ON
 * 3 function arguments for commanding GPIO ports 3, 4, 5 pins OFF
 
-If any bits are HIGH (1) in the bitmask bytes then that respective pin will be toggled ON (HIGH), if a bit is left LOW (0) then no action will take place and the bit will remain in it's current HIGH/LOW state.
+If any bits are HIGH (1) in the bitmask bytes then the respective ON/OFF action will be applied to that bitmask's pin, if a bit is left LOW (0) then no action will take place and the pin will remain in it's current HIGH/LOW state. Commanding both and HIGH and LOW for the same port/pin is not allowed.

what happens when you command both ON and OFF at the same time?

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub <https://github.com/FaradayRF/Faraday-Software/pull/111#pullrequestreview-24915964> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AAUsyK1DWdklLdm_5m94P7XuNdLRXT-_ks5rh77XgaJpZM4MLFK4> .  <https://github.com/notifications/beacon/AAUsyETMuOqejb_Z1EvCkmHmTGoUA0a-ks5rh77XgaJpZM4MLFK4.gif>

kb1lqc commented on Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:
```diff
@@ -143,15 +162,7 @@ below is a screenshot of the partial output of the tutorial script when run in a
-* [Proxy Tool - TelemetryParser](http://faraday-software.readthedocs.io/en/latest/telemetryparser.html)
-* Command Application
-
+  * Can you do this with variable length ECHO data?
```
Your example doesn't show any extra text on the echo'd string. I don't understand.

kb1lqc reviewed (approved):
Items I think need to be fixed in new Issue Ticket:
- Wording issues that make text hard to read
- Example code has long comments that make text hard to follow/scroll
- Formatting discrepancies
- Spelling errors still present
- Example output doesn't match code

I think we need to get this into the master and fix these issues on a new PR.
